ksql-engine/src/main/java/io/confluent/ksql/planner/LogicalPlanner.java

```diff
@@ -113,13 +111,37 @@ private OutputNode buildOutputNode(final PlanNode sourcePlanNode) {
         extractionPolicy,
         keyField,
         intoDataSource.getKsqlTopic(),
-        partitionByField.isPresent(),
+        partitionByField,
         analysis.getLimitClause(),
         intoDataSource.isCreate(),
         analysis.getSerdeOptions()
     );
   }
 
+  private KeyField buildOutputKeyField(
+      final PlanNode sourcePlanNode
+  ) {
+    final KeyField sourceKeyField = sourcePlanNode.getKeyField();
+
+    final Optional<String> partitionByField = analysis.getPartitionBy();
+    if (!partitionByField.isPresent()) {
+      return sourceKeyField;
+    }
+
+    final String partitionBy = partitionByField.get();
+    final LogicalSchema schema = sourcePlanNode.getSchema();
+
+    if (schema.isImplicitField(partitionBy)) {
```

Comment:
> The problem is that the keyField for a source must exist in the value schema of the source, that's the true value schema of the source, so ROWTIME or ROWKEY are not valid values for the key field. (Remember, the key field is the field in the value schema that matches the key, hence it must be in the value schema).

I'm confused. Let's assume I have `CREATE STREAM foo (foo VARCHAR) WITH(keyField='foo');` why should the result of `CREATE STREAM bar AS SELECT * FROM FOO PARTITION BY foo` be different than `CREATE STREAM bar AS SELECT * FROM FOO PARTITION BY ROWKEY`?

Reply:
They will be the same =)

In both cases  `sourceKeyField` is `KeyField(foo)`, as that's the keyField of stream `foo`.

`CREATE STREAM bar AS SELECT * FROM FOO PARTITION BY foo`

For the above  `partitionBy` will be `foo`, which won't match either `if(implicitField)` or `if(keyField)`, so this method will return `sourceKeyField.withName(partitionBy)`, i.e. `KeyField(foo)`.

`CREATE STREAM bar AS SELECT * FROM FOO PARTITION BY ROWKEY`

For the above  `partitionBy` will be `ROWKEY`, which won't match `if(implicitField)` but will match `if(keyField)`, so this method will return `sourceKeyField`, i.e. `KeyField(foo)`.

However, if the `partitionBy` is `ROWTIME`, it will match `if(implicitField)` and hence return `sourceKeyField.withName(Optional.empty())`, i.e. `KeyField.none()`.

(NB: I've ignored the legacy part of the keyField in all of this to keep it simple).

Reply:
> I think your confusion may be coming about because you think ROWKEY will match if(implicitField)

This hits the point exactly. The following code is probably why I was confused:

```java
private static final Set<String> IMPLICIT_FIELD_NAMES = ImmutableSet.of(
      SchemaUtil.ROWTIME_NAME,
      SchemaUtil.ROWKEY_NAME
  );

  private static final Schema IMPLICIT_SCHEMA = SchemaBuilder
      .struct()
      .field(SchemaUtil.ROWTIME_NAME, Schema.OPTIONAL_INT64_SCHEMA)
      .build();
```

I don't really care what we name one or the other, just so long they're not the same 😂

Reply:
Yeah, I'm planning on getting rid of the first one. It's just a temporary artefact. But I'll also rename the implicitFields to metaFields in a follow up PR.